### PR TITLE
New: Updated Gulpfile for multiple builds

### LINKED ISSUE #77: Update to support multiple builds
I think we should update to support multiple builds, as `polymer.json` has.

@FredKSchott can you help here? 😄

### PR DESCRIPTION
**Pull request by stramel**
Following closely the polymer cli's build process for multiple builds and optimizations.

- CLI flags for build command that match polymer-cli's.
    - The presets make use of any matching pre-configured preset in polymer.json
    - Generate a custom build with the other build flags
- Reads from polymer.json for builds if no CLI flags passed
- Ability to apply optimizations on only sources and/or dependencies, as well as, both sources and dependencies

Fixes #77

### CODE REVIEW

**Comment by abdonrd** [comment]
I'm not sure if we're making it too complicated...

Waiting for @FredKSchott.

**Comment by stramel** [comment]
@abdonrd I can strip out some of the complexity if you would like.

- Remove CLI args, just read from `polymer.json`?
- Remove the source and dependency splitters?
- Remove the defaulted optimizations and just provide the hook areas where the user can add theirs?

**Comment by abdonrd** [comment]
@stramel Probably we shouldn't add the CLI args. But I'm not sure.

**Comment by stramel** [comment]
Yeah, I wasn't sure... I tried to hide them at the bottom to avoid add more confusion. I could also pull some of the functionality out into another file to reduce the main gulpfile?

**Comment by abdonrd** [comment]
I think we can remove the CLI args. The `gulpfile.js` is usually for ad-hoc builds.

**Comment by stramel** [comment]
Hmmm, build isn't failing... Can someone point me in the right direction?

**Comment by abdonrd** [comment]
Ping to @FredKSchott & @justinfagnani.

File: generators/app/gulpfile.js [hunk 1]
```diff
@@ -10,56 +10,54 @@
...
+  const builds = polymerProject.config.builds || [];
 
-    // Lets create some inline code splitters in case you need them later in your build.
-    let sourcesStreamSplitter = new polymerBuild.HtmlSplitter();
-    let dependenciesStreamSplitter = new polymerBuild.HtmlSplitter();
+  // Okay, so first thing we do is clear the build directory
+  console.log(`Clearing ${mainBuildDirectoryName}${path.sep} directory...`);
+  return del([mainBuildDirectoryName])
+    .then(() => {
+      return Promise.all(builds.map((options) => {
```
Comment (justinfagnani) [inline generators/app/gulpfile.js:43]: `options` -> `buildOptions`

File: generators/app/gulpfile.js [hunk 2]
```diff
@@ -68,61 +66,215 @@ function build() {
...
+    });
+}
+gulp.task('build', runBuilds);
+
+/**
+ * Process CLI Args
+ *
+ * @return {Object}
+ */
+function getArgs() {
```
Comment (justinfagnani) [inline generators/app/gulpfile.js:196]: I would remove this, as it's starting to get into polymer-cli territory

File: generators/app/gulpfile.js [hunk 3]
```diff
@@ -10,56 +10,54 @@
...
-      .then(() => {
+        return new Promise((resolve) => {
+          console.log(`(${buildName}) Building...`);
 
-        // Let's start by getting your source files. These are all the files
-        // in your `src/` directory, or those that match your polymer.json
-        // "sources"  property if you provided one.
-        let sourcesStream = polymerProject.sources()
+          // Lets create some inline code splitters in case you need them later in your build.
+          const sourcesStreamSplitter = new polymerBuild.HtmlSplitter();
```
Comment (justinfagnani) [inline generators/app/gulpfile.js:53]: I think I would just use define splitter and add a commented out section of code showing how to wire a splitter into the sources stream.

**Comment by krumware** [comment]
With this, i have problems with concurrent builds when a build is using bundler. I'm investigating on my end, I suspect it could be file access issues (i'm on windows).

@justinfagnani: A note on the comments about being close to polymer-cli (not referring to the cmd-line args, those can go).  Personally, I like that this follows the cli closely. It lets us understand where to inject custom steps, but also be able to leverage other polymer-related projects (particularly prpl-server-node as a library, in my case)

**Comment by robdodson** [comment]
@justinfagnani Using this as a basis for something like `polymer eject` sounds like an interesting idea. When we originally created the custom build project it was for folks who wanted to add things into their build that polymer CLI didn't support out of the box (SASS, TypeScript, etc). So we were essentially recreating the `build` command but showing folks where they could hook in to add their own steps.

**Comment by stramel** [comment]
@robdodson @justinfagnani I'm just following up on this to see what we should do with this.

**Comment by robdodson** [comment]
@FredKSchott do you know if there are any plans to make the build tool support `eject`?

**Comment by justinfagnani** [comment]
@robdodson we've got our plates full with Polymer 3.0 stuff right now, so no immediate plans.

It's not particularly hard work though. First, we need to move some transforms from polymer-cil to polymer-build, then we just need some sort of Yeoman template for the gulpfile.

If anyone's interested in picking up this work it'd be a great contribution. We could document the requirements to help it along.

**Comment by stramel** [comment]
@justinfagnani I would be interested in taking the lead on getting this running. If you can outline the requirements and maybe the transforms to move, then I will get everything going.

**Comment by justinfagnani** [comment]
@stramel Awesome! This is the issue on the CLI here: https://github.com/Polymer/polymer-cli/issues/573 let's take the discussion there.

**Comment by stramel** [comment]
As @justinfagnani stated above, we are taking the conversation and effort to making a polymer cli command to cover this.

I will leave my branch out here for anyone pointing to it for now.

**Pull request closed without merging**